contrib/log/slog/slog.go

```diff
@@ -31,21 +39,71 @@ func NewJSONHandler(w io.Writer, opts *slog.HandlerOptions) slog.Handler {
 
 // WrapHandler enhances the given logger handler attaching tracing information to logs.
 func WrapHandler(h slog.Handler) slog.Handler {
-	return &handler{h}
+	return &handler{wrapped: h}
 }
 
 type handler struct {
-	slog.Handler
+	wrapped slog.Handler
+	groups  []group
+}
+
+// Enabled calls the wrapped handler Enabled method.
+func (h *handler) Enabled(ctx context.Context, level slog.Level) bool {
+	return h.wrapped.Enabled(ctx, level)
 }
 
 // Handle handles the given Record, attaching tracing information if found.
 func (h *handler) Handle(ctx context.Context, rec slog.Record) error {
+	reqHandler := h.wrapped
+
+	// We need to ensure the trace id and span id keys are set at the root level:
+	// https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/
+	// In case the user has created group loggers, we ignore those and
+	// set them at the root level.
 	span, ok := tracer.SpanFromContext(ctx)
 	if ok {
-		rec.Add(
-			slog.Uint64(ext.LogKeyTraceID, span.Context().TraceID()),
-			slog.Uint64(ext.LogKeySpanID, span.Context().SpanID()),
-		)
+		traceID := strconv.FormatUint(span.Context().TraceID(), 10)
+		spanID := strconv.FormatUint(span.Context().SpanID(), 10)
+
+		attrs := []slog.Attr{
+			slog.String(ext.LogKeyTraceID, traceID),
+			slog.String(ext.LogKeySpanID, spanID),
+		}
+		reqHandler = reqHandler.WithAttrs(attrs)
+	}
+	for _, g := range h.groups {
+		reqHandler = reqHandler.WithGroup(g.name)
+		if len(g.attrs) > 0 {
+			reqHandler = reqHandler.WithAttrs(g.attrs)
+		}
+	}
+	return reqHandler.Handle(ctx, rec)
+}
+
+// WithAttrs saves the provided attributes associated to the current Group.
+// If Group was not called for the logger, we just call WithAttrs for the wrapped handler.
+func (h *handler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	if len(h.groups) == 0 {
+		return &handler{
+			wrapped: h.wrapped.WithAttrs(attrs),
+			groups:  h.groups,
+		}
+	}
+	groups := append([]group{}, h.groups...)
+	curGroup := groups[len(groups)-1]
+	curGroup.attrs = append(curGroup.attrs, attrs...)
+	groups[len(groups)-1] = curGroup
```

Comment:
So a call to `WithAttrs` applies attributes to the most-recently added group. I'm assuming `WithAttrs` is only ever called when initializing the Handler, like this?

`Handler.WithAttrs(<attr1>).WithGroup(<group1>).WithAttrs(<attr2>)`

(Where attr1 is applied globally to the handler, and attr2 is applied to group1?)

If you were to use it like this:

```
handler := Handler.WithGroup(<group1>)
<some other stuff>
handler = handler.WithGroup(<group2>)
<some other stuff>
handler = handler.WithAttrs(<attr1>)
```

Then it's less obvious that attr1 would be applied to only group2. 

So just confirming this is all in-line with how slog is used. 👍

Reply:
The handler methods are not called directly, these are used through the `slog` [api](https://pkg.go.dev/log/slog#hdr-Groups) like:

```
logger := slog.Default().With("id", systemID).WithGroup("group").With("nested", "attr")
```

Feel free to check the slog documentation to confirm, but basically what you said is correct 👍